Rewrite opentherm_gw to a component

mvn23 opened the pull request:
## Description:
Rewrite opentherm_gw to a component. This will allow us to expose internal data as sensors.

**Related issue (if applicable):** ref. #16670

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#6637

## Example entry for `configuration.yaml` (if applicable):
```yaml
opentherm_gw:
  device: /dev/ttyUSB0
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - N/A

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

mvn23 committed:
Rewrite opentherm_gw to a component which loads the opentherm_gw climate platform.

mvn23 committed:
Add OpenTherm Gateway sensor platform.

* Remove library imports from platforms (use hass.data instead)
* Update .coveragerc
* Update docstrings to use new component documentation url

mvn23 committed:
Add OpenTherm Gateway binary sensor support.
Fix houndci findings.

MartinHjelmare commented:
Please only do the conversion to component in this PR. Add the new platforms in separate PRs, one platform and PR at a time.

mvn23 committed:
Revert "Add OpenTherm Gateway binary sensor support."

This reverts commit 5711dc4c25edc7352ba8bdf14875c79f0bf51d11.

mvn23 committed:
Revert "Add OpenTherm Gateway sensor platform."

This reverts commit b3505ed5611d52842c1ea623cd04a13ef2bb9a88.

mvn23 committed:
Remove import from platform, use hass.data instead.
Update .coveragerc
Update docstrings
Update requirements_all.txt
General code cleanup

MartinHjelmare commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -1,34 +1,23 @@
-For more details about this component, please refer to the documentation at
-http://home-assistant.io/components/climate.opentherm_gw/
+For more details about this platform, please refer to the documentation at
+http://home-assistant.io/components/opentherm_gw/
```
The url is wrong. Revert to old url. Each platform should have its own docs page.

MartinHjelmare commented on homeassistant/components/opentherm_gw.py:
```diff
@@ -0,0 +1,76 @@
+http://home-assistant.io/components/opentherm_gw/
+"""
+import logging
+import voluptuous as vol
```
Add a blank line between standard library and 3rd party imports.

MartinHjelmare commented on homeassistant/components/opentherm_gw.py:
```diff
@@ -0,0 +1,76 @@
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_DEVICE): cv.string,
+        vol.Optional(CONF_CLIMATE, default=CLIMATE_SCHEMA({})):
```
Just pass an empty dict as default value. It will be automatically passed to the schema.

MartinHjelmare commented on homeassistant/components/opentherm_gw.py:
```diff
@@ -0,0 +1,76 @@
+async def async_setup(hass, config):
+    """Set up the OpenTherm Gateway component."""
+    conf = config.get(DOMAIN)
+    if conf is None:
```
This can't happen.

MartinHjelmare commented on homeassistant/components/opentherm_gw.py:
```diff
@@ -0,0 +1,76 @@
+
+async def async_setup(hass, config):
+    """Set up the OpenTherm Gateway component."""
+    conf = config.get(DOMAIN)
```
Use `dict[key]`.

MartinHjelmare commented on homeassistant/components/opentherm_gw.py:
```diff
@@ -0,0 +1,76 @@
+        DATA_DEVICE: pyotgw.pyotgw(),
+        DATA_GW_VARS: pyotgw.vars,
+    }
+    hass.async_add_job(connect_and_subscribe, hass, conf)
```
Use `hass.async_create_task`.

mvn23 committed:
Fix review findings.
Avoid using hass.data within connect_and_subscribe.

mvn23 commented:
Thanks for the review, all issues should be addressed.

MartinHjelmare reviewed (approved):
Looks good! Can be merged when docs PR is linked in the description and build passes.

frenck commented:
Docs appeared! Thanks, @mvn23 👍
